### DISCUSSION TITLE
Better Modules Naming Scheme

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#13775] [Ideas]
Currently, you can specify a module naming scheme like:
```yaml
naming_scheme: '{name}/{version}'
```
I would like the ability to change this to:
```yaml
naming_scheme: '{name}/{version}/{hash:7}'
```
Where `{hash}` is the hash and `7` is the hash length.

### Rationale

When I load a module, I have several options. If I don't care which version, I can say:
```console
$ module load zlib
```
If I do care about the version, I would like to be able to do:
```console
$ module load zlib/1.2.8
```
Currently this doesn't work, because Spack automatically appends the hash after a dash instead of a slash.

### Description

Currently, I've been using the following workaround locally:
```diff
diff --git a/lib/spack/spack/modules/common.py b/lib/spack/spack/modules/common.py
index 434d09c12..e6c06bdd7 100644
--- a/lib/spack/spack/modules/common.py
+++ b/lib/spack/spack/modules/common.py
@@ -538,7 +538,7 @@ class BaseFileLayout(object):
         name = os.path.join(*parts)
         # Add optional suffixes based on constraints
         path_elements = [name] + self.conf.suffixes
-        return '-'.join(path_elements)
+        return '/'.join(path_elements)
 
     @property
     def filename(self):
```
By making this a configuration option, users can have more flexibility as to how module files are named.

@becker33 @alalazo

**Comment by albestro** [comment]
Just yesterday I was about opening an issue about a module names conflict I have with my package set.

I have the same package installed with different variants.

e.g.
```bash
wyna7zj hpx@1.3.0 build_type=Debug ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=jemalloc max_cpu_count=128 ~networking~tools
zya72yn hpx@1.3.0 build_type=Debug ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=tcmalloc max_cpu_count=128 ~networking~tools
mn4ukz7 hpx@1.3.0 build_type=Release ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=jemalloc max_cpu_count=128 ~networking~tools
ha562lm hpx@1.3.0 build_type=Release ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=tcmalloc max_cpu_count=128 ~networking~tools
```

I cannot even create the modules because this result in a name conflicts:

```bash
spack module tcl refresh hpx
==> You are about to regenerate tcl module files for:

-- cray-cnl7-broadwell / gcc@8.3.0 ------------------------------
wyna7zj hpx@1.3.0  zya72yn hpx@1.3.0  mn4ukz7 hpx@1.3.0  ha562lm hpx@1.3.0

==> Do you want to proceed? [y/n] y
==> Error: Name clashes detected in module files:

file: /apps/daint/SSL/software/spack-current/share/spack/modules/cray-cnl7-broadwell/hpx/1.3.0-gcc-8.3.0-cray-cnl7-broadwell-mpi-python3
spec: hpx@1.3.0%gcc@8.3.0 build_type=Debug ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=jemalloc max_cpu_count=128 ~networking~tools arch=cray-cnl7-broadwell
spec: hpx@1.3.0%gcc@8.3.0 build_type=Debug ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=tcmalloc max_cpu_count=128 ~networking~tools arch=cray-cnl7-broadwell

file: /apps/daint/SSL/software/spack-current/share/spack/modules/cray-cnl7-broadwell/hpx/1.3.0-gcc-8.3.0-cray-cnl7-broadwell-release-mpi-python3
spec: hpx@1.3.0%gcc@8.3.0 build_type=Release ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=jemalloc max_cpu_count=128 ~networking~tools arch=cray-cnl7-broadwell
spec: hpx@1.3.0%gcc@8.3.0 build_type=Release ~cuda cuda_arch=none cxxstd=14 instrumentation=none malloc=tcmalloc max_cpu_count=128 ~networking~tools arch=cray-cnl7-broadwell

==> Error: Operation aborted
```

I think this may be a good support for the @adamjstewart's feature request.

**Comment by alalazo** [comment]
@adamjstewart This is not a big change, but what is it giving more than:
```yaml
modules:
  tcl:
    hash_length: 0
```
plus the possibility to add suffixes? If you are in a situation like:
```console
-- cray-cnl7-broadwell / gcc@8.3.0 ------------------------------
wyna7zj hpx@1.3.0  zya72yn hpx@1.3.0  mn4ukz7 hpx@1.3.0  ha562lm hpx@1.3.0
```
would:
```console
$ spack load hpx@1.3.0
```
load anything deterministically in your case? @albestro How do you envision loading each one of the four `hpx` above under the scheme proposed here AND without using the hash?

**Comment by albestro** [comment]
I never tried, but I would make the spec more detailed.

```console
spack load hpx@1.3.0 build_type=Release malloc=jemalloc
```

Would it be possible?

**Comment by alalazo** [comment]
@albestro If you do it like that (i.e. you use `spack load`) you don't need to remove the hash in the first place when you generate module files. @adamjstewart was referring to a case where `module load` was used instead of `spack load`.

**Comment by albestro** [comment]
Yes, true.

The problem is that at the moment I cannot generate modules for these packages at all. But, I thought that what @adamjstewart was suggesting in the end of his first comment

>By making this a configuration option, users can have more flexibility as to how module files are named.

could solve my problem. Isn't it?

**Comment by alalazo** [comment]
@albestro My point is that you can probably solve your issue right now by changing your configuration. Can you share the output of:
```console
$ spack config blame modules
```
?

**Comment by albestro** [comment]
You pointed me in the right direction, I didn't know that I was able to customize the module name
```
modules:
  tcl:
    naming_scheme: '{name}/{version}-{compiler.name}-{compiler.version}-{architecture.target}-{hash:7}'
```

Thank you for helping me!

**Comment by adamjstewart** [comment]
@alalazo Yes, this is specific to `module load`, I’ve never used `spack load`. I never remove hashes or bother with suffixes as Spack tends to install dozens of copies of the same package without me asking for it.

**Comment by matz-e** [comment]
We had done this local modification in our deployed software installation on our main cluster in production, @adamjstewart.

I don't think it served as that well in the end, since we conditionally appended stuff like `/parallel`. E.g., `hdf5/1.10.5/parallel` for `hdf5+mpi`. But with tcl modules, that barred us from just having `hdf5/1.10.5` for a non-MPI version, since that already is a directory. This essentially locked us into a undesirable directory structure, as users relied on the naming scheme as we implemented it…

So, I would keep the `-` to join suffixes…
